File: src/domain/garment-detail.test.ts
import { describe, expect, it } from 'vitest';
import { garmentDetail } from './garment-detail';
import type { Garment } from '../types';

const garment = (overrides: Partial<Garment>): Garment => ({
  id: 'g',
  image_uri: 'front.jpg',
  image_uri_nobg: null,
  image_uris: ['front.jpg'],
  image_uris_nobg: [],
  category: 'tops',
  subcategory: 'T-Shirt',
  subcategories: ['T-Shirt'],
  tags: [],
  brand: null,
  color_primary: '#000000',
  color_secondary: null,
  color_palette: ['#000000'],
  size: 'M',
  purchase_date: null,
  is_available: true,
  unavailable_date: null,
  created_at: '2026-01-01T00:00:00.000Z',
  updated_at: '2026-01-01T00:00:00.000Z',
  ...overrides,
});

describe('which photo is shown', () => {
  it('prefers the cut-out for the selected photo', () => {
    const view = garmentDetail(
      garment({ image_uris: ['a.jpg', 'b.jpg'], image_uris_nobg: ['', 'b-cut.png'] }),
      1
    );

    expect(view.displayedImage).toBe('b-cut.png');
    expect(view.gallery.map(g => g.uri)).toEqual(['a.jpg', 'b-cut.png']);
  });

  it('falls back to the original when that photo has no cut-out', () => {
    const view = garmentDetail(
      garment({ image_uris: ['a.jpg', 'b.jpg'], image_uris_nobg: ['a-cut.png', ''] }),
      1
    );

    expect(view.displayedImage).toBe('b.jpg');
  });

  it('has nothing to show for a garment with no photos', () => {
    const view = garmentDetail(garment({ image_uri: '', image_uris: [], image_uris_nobg: [] }));

    expect(view.displayedImage).toBeNull();
    expect(view.gallery).toEqual([]);
    expect(view.showsGallery).toBe(false);
  });

  it('offers the strip only when there is a choice', () => {
    expect(garmentDetail(garment({ image_uris: ['a.jpg'] })).showsGallery).toBe(false);
    expect(garmentDetail(garment({ image_uris: ['a.jpg', 'b.jpg'] })).showsGallery).toBe(true);
  });

  it('clamps a selection past the end rather than showing nothing selected', () => {
    // A remembered index, or a garment whose photos were edited since. Reading
    // past the end used to fall through to the first photo while the strip
    // showed none of them selected, so the screen disagreed with itself.
    const view = garmentDetail(garment({ image_uris: ['a.jpg', 'b.jpg'] }), 7);

    expect(view.selectedIndex).toBe(0);
    expect(view.displayedImage).toBe('a.jpg');
    expect(view.gallery.map(g => g.selected)).toEqual([true, false]);
  });

  it('clamps a negative selection', () => {
    const view = garmentDetail(garment({ image_uris: ['a.jpg', 'b.jpg'] }), -1);

    expect(view.selectedIndex).toBe(0);
    expect(view.gallery.map(g => g.selected)).toEqual([true, false]);
  });

  it('marks which photos have a cut-out', () => {
    const view = garmentDetail(
      garment({ image_uris: ['a.jpg', 'b.jpg'], image_uris_nobg: ['a-cut.png', ''] })
    );

    expect(view.gallery.map(g => g.hasCutout)).toEqual([true, false]);
  });
});

describe('the background-removal button', () => {
  it('offers removal for a photo that has no cut-out', () => {
    expect(garmentDetail(garment({ image_uris: ['a.jpg'] })).backgroundAction).toBe('remove');
  });

  it('offers undo only while a separate original still exists', () => {
    const legacy = garment({ image_uris: ['a.jpg'], image_uris_nobg: ['a-cut.png'] });

    expect(garmentDetail(legacy).backgroundAction).toBe('undo');
  });

  it('offers nothing once the cut-out has replaced the original', () => {
    // Removal now overwrites the photo it came from, so there is nothing to
    // revert to -- and an undo button here would destroy the only copy.
    const replaced = garment({ image_uris: ['a-cut.png'], image_uris_nobg: ['a-cut.png'] });

    expect(garmentDetail(replaced).backgroundAction).toBeNull();
  });
});

describe('the palette', () => {
  it('names the colours that came from the picker', () => {
    const view = garmentDetail(garment({ color_palette: ['#000000', '#0066CC'] }));

    expect(view.palette).toEqual([
      { hex: '#000000', colorKey: 'black' },
      { hex: '#0066CC', colorKey: 'blue' },
    ]);
  });

  it('names a colour whatever case it was stored in', () => {
    // The wardrobe holds both cases -- the analytics grouping had to be fixed
    // for the same reason -- and '#cc0000' is the same red as '#CC0000'.
    const view = garmentDetail(garment({ color_palette: ['#cc0000'] }));

    expect(view.palette[0].colorKey).toBe('red');
  });

  it('leaves a colour that was not picked from the palette unnamed', () => {
    const view = garmentDetail(garment({ color_palette: ['#123456'] }));

    expect(view.palette).toEqual([{ hex: '#123456', colorKey: null }]);
  });

  it('names the multi-colour sentinel', () => {
    expect(garmentDetail(garment({ color_palette: ['#RAINBOW'] })).palette[0].colorKey).toBe('multi');
  });

  it('falls back to the primary and secondary columns', () => {
    const view = garmentDetail(
      garment({ color_palette: [], color_primary: '#FFFFFF', color_secondary: '#000080' })
    );

    expect(view.palette.map(p => p.colorKey)).toEqual(['white', 'navy']);
  });
});

describe('tags', () => {
  it('takes the seasons out of the tags', () => {
    const view = garmentDetail(garment({ tags: ['cotton', 'summer', 'striped'] }));

    expect(view.tags).toEqual(['cotton', 'striped']);
    expect(view.seasons).toEqual(['summer']);
  });

  it('drops tags that used to be structured filters', () => {
    // Old rows still carry them, and a restored backup reintroduces them, so
    // they must not surface as if the user had typed them.
    const view = garmentDetail(garment({ tags: ['formal', 'rainy', 'linen'] }));

    expect(view.tags).toEqual(['linen']);
  });

  it('lists seasons in the app order rather than the order they were typed', () => {
    const view = garmentDetail(garment({ tags: ['winter', 'spring', 'winter'] }));

    expect(view.seasons).toEqual(['spring', 'winter']);
  });
});

describe('the fields with nothing in them', () => {
  it('treats a blank brand or size as absent', () => {
    const view = garmentDetail(garment({ brand: '   ', size: '' }));

    expect(view.brand).toBeNull();
    expect(view.size).toBeNull();
  });

  it('trims a brand that has something in it', () => {
    expect(garmentDetail(garment({ brand: ' Nike ' })).brand).toBe('Nike');
  });

  it('reports an unavailable date only for a garment that is unavailable', () => {
    const gone = garmentDetail(
      garment({ is_available: false, unavailable_date: '2026-03-04T00:00:00.000Z' })
    );
    // A row can keep the date from a previous spell of being unavailable.
    const back = garmentDetail(
      garment({ is_available: true, unavailable_date: '2026-03-04T00:00:00.000Z' })
    );

    expect(gone.unavailableDate).toBe('2026-03-04T00:00:00.000Z');
    expect(back.unavailableDate).toBeNull();
  });
});

describe('occasions', () => {
  it('derives them from the garment type', () => {
    expect(garmentDetail(garment({ subcategories: ['Blazer'] })).occasions)
      .toEqual(['work', 'formal']);
  });

  it('falls back for a type it does not recognise', () => {
    expect(garmentDetail(garment({ category: 'loungewear', subcategories: ['Unknown'] })).occasions)
      .toEqual(['lounge']);
  });
});

File: presentation/src/test/kotlin/com/wardrobapp/presentation/GarmentDetailTest.kt
package com.wardrobapp.presentation

import com.wardrobapp.data.normalizeGarmentRow
import com.wardrobapp.domain.Occasion
import com.wardrobapp.domain.Season
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * What a garment's detail screen shows.
 *
 * A view function over a database row, and most of its decisions are about
 * disagreeing with itself: a photo strip whose selection is not the photo on
 * screen, a colour swatch with no name under it, a cut-out shown where the
 * original was expected.
 */
class GarmentDetailTest {

    private fun record(
        images: String = """["front.jpg"]""",
        cutouts: String = "[]",
        category: String = "tops",
        subcategories: String = """["T-Shirt"]""",
        tags: String = "[]",
        brand: String? = null,
        size: String? = null,
        palette: String = """["#1F3A93"]""",
        isAvailable: Int = 1,
        unavailableDate: String? = null,
        purchaseDate: String? = null,
    ) = normalizeGarmentRow(
        mapOf(
            "id" to "g1",
            // normalizeGarmentRow folds the single-value columns into the list
            // ones, so these are the first entry rather than another photo.
            "image_uri" to (Regex("\"([^\"]+)\"").find(images)?.groupValues?.get(1) ?: ""),
            "image_uris" to images,
            "image_uris_nobg" to cutouts,
            "category" to category,
            "subcategories" to subcategories,
            "tags" to tags,
            "brand" to brand,
            "size" to size,
            "color_primary" to (Regex("\"([^\"]+)\"").find(palette)?.groupValues?.get(1) ?: ""),
            "color_palette" to palette,
            "is_available" to isAvailable,
            "unavailable_date" to unavailableDate,
            "purchase_date" to purchaseDate,
        ),
        "",
    )

    @Test
    fun `one photo is shown without a strip to choose from`() {
        val view = garmentDetail(record())

        assertEquals("front.jpg", view.displayedImage)
        assertEquals(1, view.gallery.size)
        assertTrue(!view.showsGallery, "a strip was offered for a single photo")
        assertEquals(0, view.selectedIndex)
    }

    @Test
    fun `the photo on screen is the one the strip says is selected`() {
        val view = garmentDetail(
            record(images = """["a.jpg","b.jpg","c.jpg"]"""),
            selectedIndex = 1,
        )

        assertEquals("b.jpg", view.displayedImage)
        assertEquals(1, view.selectedIndex)
        assertEquals(listOf(false, true, false), view.gallery.map { it.selected })
        assertTrue(view.showsGallery)
    }

    @Test
    fun `a selection from outside the strip falls back to the first photo, visibly`() {
        // A remembered index, or a garment whose photos were edited since. Reading
        // past the end used to show the first photo while the strip showed nothing
        // selected, so the screen disagreed with itself.
        for (index in listOf(-1, 3, 99)) {
            val view = garmentDetail(record(images = """["a.jpg","b.jpg"]"""), selectedIndex = index)

            assertEquals("a.jpg", view.displayedImage, "index $index")
            assertEquals(0, view.selectedIndex, "index $index")
            assertEquals(listOf(true, false), view.gallery.map { it.selected }, "index $index")
        }
    }

    @Test
    fun `a cut-out is shown in place of the photo it was cut from`() {
        val view = garmentDetail(
            record(images = """["a.jpg","b.jpg"]""", cutouts = """["a-cut.png",""]"""),
        )

        assertEquals("a-cut.png", view.displayedImage)
        assertEquals(listOf("a-cut.png", "b.jpg"), view.gallery.map { it.uri })
        assertEquals(listOf(true, false), view.gallery.map { it.hasCutout })
    }

    @Test
    fun `a garment with no photos at all does not crash`() {
        val view = garmentDetail(record(images = "[]"))

        assertEquals(null, view.displayedImage)
        assertEquals(emptyList(), view.gallery)
        assertTrue(!view.showsGallery)
    }

    @Test
    fun `seasons read in the app's own order, and the rest stay as tags`() {
        // Season tags and free tags share one column, so splitting them is what
        // stops "winter" showing up as something the user typed.
        val view = garmentDetail(record(tags = """["winter","favourite","spring"]"""))

        assertEquals(listOf(Season.SPRING, Season.WINTER), view.seasons)
        assertEquals(listOf("favourite"), view.tags)
    }

    @Test
    fun `occasions come from what the garment is, not from what it was tagged`() {
        val view = garmentDetail(record(category = "tops", subcategories = """["Blouse"]"""))

        assertEquals(listOf(Occasion.WORK, Occasion.FORMAL), view.occasions)
    }

    @Test
    fun `a blank field is absent rather than an empty line`() {
        val view = garmentDetail(record(brand = "   ", size = "", purchaseDate = null))

        assertEquals(null, view.brand)
        assertEquals(null, view.size)
        assertEquals(null, view.purchaseDate)
    }

    @Test
    fun `a retired garment says when, and an available one says nothing`() {
        val retired = garmentDetail(
            record(isAvailable = 0, unavailableDate = "2026-02-01T00:00:00.000Z")
        )
        assertTrue(!retired.isAvailable)
        assertEquals("2026-02-01T00:00:00.000Z", retired.unavailableDate)

        // A date left behind by an earlier retirement must not show on a garment
        // that is back in use.
        val restored = garmentDetail(
            record(isAvailable = 1, unavailableDate = "2026-02-01T00:00:00.000Z")
        )
        assertTrue(restored.isAvailable)
        assertEquals(null, restored.unavailableDate)
    }

    @Test
    fun `the background button offers the action that is available`() {
        val original = garmentDetail(record(images = """["a.jpg"]""", cutouts = "[]"))
        val cutout = garmentDetail(record(images = """["a.jpg"]""", cutouts = """["a-cut.png"]"""))

        assertEquals(BackgroundAction.REMOVE, original.backgroundAction)
        assertEquals(BackgroundAction.UNDO, cutout.backgroundAction)

        // A garment imported as a cut-out has the same path in both slots, so
        // there is nothing to undo to and no original to run removal against --
        // which is the case the form used to get wrong, offering undo where
        // there was nothing to undo.
        val importedCutout = garmentDetail(
            record(images = """["cut.png"]""", cutouts = """["cut.png"]""")
        )
        assertEquals(null, importedCutout.backgroundAction)
    }

    @Test
    fun `every colour the picker offers has a name`() {
        // A hex in the picker that this cannot name would show as a raw code on
        // the detail screen -- the failure mode is silent, so it is stated here.
        val duplicates = GARMENT_COLORS.groupBy { it.second.uppercase() }.filterValues { it.size > 1 }
        assertEquals(emptyMap(), duplicates, "two colour keys share a hex")

        for ((key, hex) in GARMENT_COLORS) {
            val named = garmentDetail(record(palette = """["$hex"]""")).palette.single().colorKey
            assertEquals(key, named, "$hex is not named")
        }
    }

    @Test
    fun `a colour the picker does not have is shown without a name rather than dropped`() {
        val view = garmentDetail(record(palette = """["#123456"]"""))

        assertEquals("#123456", view.palette.single().hex)
        assertEquals(null, view.palette.single().colorKey)
    }
}
